### PR TITLE
Add Gemma4 MaxText→vLLM Weight Converter (torchax)

### PR DESCRIPTION
**Pull request by hengtaoguo**
# Description

Original author: @khatwanimohit @aireenmei https://github.com/AI-Hypercomputer/maxtext/pull/3677
Refactor/test: @hengtaoguo 

Extracts the Gemma4-specific weight conversion logic from `bench_weight_sync.py` into a proper converter class, following the same base/model-specific split established for Qwen3.

**New: `gemma4_moe.py`**
- Adds `Gemma4MaxTextToVLLMConverter`, inheriting from `BaseMaxTextToVLLMConverter`
- Supports `gemma4-26b` (MoE: 128 routed + 1 shared expert)
- Handles Gemma4's scanned-block layout (6 slots × N reps, local + global attention)
- Overrides `convert()` to add the `_convert_norms` step and dispatch MoE vs. dense MLP

**Updated: `validate_converter.py`**
- Imports `Gemma4MaxTextToVLLMConverter` and dispatches on `gemma4-*` model names
- Adds `gemma4-26b` entry to `vllm_model_name_mapping`

**Notes**: 
1. Set env var `MODEL_IMPL_TYPE=vllm` to force the torchax-backed vLLM model for Gemma4 (default `"auto"` resolves to `"flax_nnx"` in newer tpu-inference, which uses a nested Flax state incompatible with the flat-key converter output)
2. Gemma4's prompt need to start with `<bos>`, example: `prompt="<box>Paris is"`

# Tests

Tested with `validate_converter`, [full logs](https://paste.googleplex.com/5016985286082560):

```
export MODEL_IMPL_TYPE="vllm"
python -m maxtext.integration.vllm.torchax_converter.validate_converter src/maxtext/configs/base.yml model_name=gemma4-26b tokenizer_type=huggingface tokenizer_path=google/gemma-4-26b-a4b-it load_parameters_path=gs://maxtext-gemma/gemma4/26b/converted/2026-04-07-23-04/0/items run_name=gemma4_converter_validation per_device_batch_size=1 max_prefill_predict_length=8 max_target_length=16 steps=1 scan_layers=true skip_jax_distributed_system=true weight_dtype=bfloat16 attention=dot_product remat_policy=custom decoder_layer_input=offload query_proj=offload key_proj=offload value_proj=offload ici_expert_parallelism=4 rollout_tensor_parallelism=4 hbm_utilization_vllm=0.8 async_scheduling=false prompt=\<bos\>Paris\ is hf_access_token=xxx 

# Also works with pathways ckpt at gs://hengtaoguo-maxtext-logs/checkpoints/gemma4-26b/scanned/gemma-4-26B-A4B-pathways/0/items
# python -m maxtext.checkpoint_conversion.to_maxtext src/maxtext/configs/base.yml model_name=gemma4-26b base_output_directory=gs://hengtaoguo-maxtext-logs/checkpoints/gemma4-26b/scanned/gemma-4-26B-A4B-pathways scan_layers=true hf_access_token=xxx weight_dtype=bfloat16 hardware=cpu skip_jax_distributed_system=True checkpoint_storage_use_ocdbt=False checkpoint_storage_use_zarr3=False
```

```
[RequestOutput(request_id=0, prompt='<bos>Paris is', prompt_token_ids=[2, 2, 50429, 563], encoder_prompt=None, encoder_prompt_token_ids=None, prompt_logprobs=None, outputs=[CompletionOutput(index=0, text=' the capital of France. It is accommodated in the', token_ids=[506, 5279, 529, 7001, 236761, 1030, 563, 98509, 528, 506], routed_experts=None, cumulative_logprob=None, logprobs=None, finish_reason=length, stop_reason=None)], finished=True, metrics=None, lora_request=None, num_cached_tokens=0)]
```

Tuned checkpoint with chat template `use_chat_template=True`:
```
export MODEL_IMPL_TYPE="vllm"
python -m maxtext.integration.vllm.torchax_converter.validate_converter src/maxtext/configs/base.yml model_name=gemma4-26b tokenizer_type=huggingface tokenizer_path=google/gemma-4-26b-a4b-it load_parameters_path=gs://hengtaoguo-maxtext-logs/checkpoints/gemma4-26b/scanned/gemma-4-26B-A4B-it/0/items run_name=gemma4_converter_validation per_device_batch_size=1 max_prefill_predict_length=8 max_target_length=32 steps=1 scan_layers=true skip_jax_distributed_system=true weight_dtype=bfloat16 attention=dot_product remat_policy=custom decoder_layer_input=offload query_proj=offload key_proj=offload value_proj=offload ici_expert_parallelism=4 rollout_tensor_parallelism=4 hbm_utilization_vllm=0.8 async_scheduling=false prompt=\<bos\>Paris\ is hf_access_token=xxx use_chat_template=True
```

```
[RequestOutput(request_id=0, prompt='<bos><|turn>user\nWhere is Paris?<turn|>\n<|turn>model\n<|channel>thought\n<channel|>', prompt_token_ids=[2, 2, 105, 2364, 107, 10936, 563, 9079, 236881, 106, 107, 105, 4368, 107, 100, 45518, 107, 101], encoder_prompt=None, encoder_prompt_token_ids=None, prompt_logprobs=None, outputs=[CompletionOutput(index=0, text='"Paris" can-refer-to\'s own context depends on', token_ids=[236775, 50429, 236775, 740, 236772, 35380, 236772, 1071, 236789, 236751, 1852, 4403, 9796, 580], routed_experts=None, cumulative_logprob=None, logprobs=None, finish_reason=length, stop_reason=None)], finished=True, metrics=None, lora_request=None, num_cached_tokens=0)]
```

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/integration/vllm/torchax_converter/validate_converter.py [hunk 1]
```diff
@@ -62,6 +63,7 @@
     "qwen3-30b-a3b": "Qwen/Qwen3-30B-A3B",
     "qwen3-30b-a3b-base": "Qwen/Qwen3-30B-A3B",
     "qwen3-235b-a22b": "Qwen/Qwen3-235B-A22B",
+    "gemma4-26b": "google/gemma-4-26B-A4B-it",
```
Comment (aireenmei) [inline src/maxtext/integration/vllm/torchax_converter/validate_converter.py]: I thought we want to support the non-tuned one without "-it"?

File: src/maxtext/integration/vllm/torchax_converter/validate_converter.py [hunk 2]
```diff
@@ -131,9 +145,12 @@ def validate_converter(config) -> None:
       tensor_parallel_size=config.rollout_tensor_parallelism,
       gpu_memory_utilization=getattr(config, "hbm_utilization_vllm", 0.5),
       async_scheduling=getattr(config, "async_scheduling", False),
+      # load_format="dummy",  # Load actual weights instead of dummy for debugging
```
[the remark this answers is not in the document]
Comment (aireenmei) [reply src/maxtext/integration/vllm/torchax_converter/validate_converter.py]: load_format="dummy" was not working for me with torchax

File: src/maxtext/integration/vllm/torchax_converter/validate_converter.py [hunk 3]
```diff
@@ -116,7 +127,10 @@ def validate_converter(config) -> None:
       logging.info("Name: %s, shape: %s", path_str, leaf.shape)
       logging.info("\tSharding: %s", leaf.sharding)
 
-  converter = Qwen3MaxTextToVLLMConverter(config, mesh)
+  if config.model_name.startswith("gemma4"):
```
Comment (aireenmei) [inline src/maxtext/integration/vllm/torchax_converter/validate_converter.py:131]: When "gemma4", how about we add `<bos>` to the beginning of prompt if it doesn't have

**Review by aireenmei — APPROVED** [review]

File: src/maxtext/integration/vllm/torchax_converter/validate_converter.py (hunk 3 above)
Comment (hengtaoguo) [reply src/maxtext/integration/vllm/torchax_converter/validate_converter.py:131]: Great suggestions, added a `<bos>` check and append if it doesn't exist in the Gemma4 prompt.

File: src/maxtext/integration/vllm/torchax_converter/validate_converter.py (hunk 1 above)
Comment (hengtaoguo) [reply src/maxtext/integration/vllm/torchax_converter/validate_converter.py]: SG, removed `-it` since RL should start with base model.

File: src/maxtext/integration/vllm/torchax_converter/validate_converter.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (hengtaoguo) [reply src/maxtext/integration/vllm/torchax_converter/validate_converter.py]: Same for me. It works with Qwen3 but not Gemma4. I'm commenting out this line for debugging purposes.

File: src/maxtext/integration/vllm/torchax_converter/gemma4_moe.py [hunk 4]
```diff
@@ -0,0 +1,414 @@
...
+                w2_weight_scale=None,
+                w2_bias=None,
+            ),
+            moe_backend=MoEBackend.GMM_TP,
+            w13_reorder_size=self.vllm_tp,
+            w13_interleave=False,  # Gemma4 uses gelu, not swigluoai
+        )
+        self.vllm_state[f"{p}.moe.experts.w13_weight"] = processed.w13_weight
+        self.vllm_state[f"{p}.moe.experts.w2_weight"] = processed.w2_weight
+        # Shared expert: gate+up fused, TP-interleaved (MergedColumnParallelLinear,
```
[the remark this answers is not in the document]
Comment (hengtaoguo) [reply src/maxtext/integration/vllm/torchax_converter/gemma4_moe.py:362]: Fixed

File: src/maxtext/integration/vllm/torchax_converter/gemma4_moe.py [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+      gate_up_proj = jnp.unstack(gate_up, axis=0)
+
+      # Down proj: (E, L, d_inner, d_model) -> L × (E, d_model, d_inner)
+      down_proj = jnp.unstack(jnp.transpose(routed["wo"], (1, 0, 3, 2)), axis=0)
+
+      # Shared expert:
+      #   wi_0/wi_1: (d_model, L, d_sh) -> L × (d_sh, d_model)
+      #   wo:        (d_sh, L, d_model)  -> L × (d_model, d_sh)
+      sh_gate = jnp.unstack(jnp.transpose(shared["wi_0"]["kernel"], (1, 2, 0)), axis=0)
+      sh_up = jnp.unstack(jnp.transpose(shared["wi_1"]["kernel"], (1, 2, 0)), axis=0)
```
[the remark this answers is not in the document]
Comment (hengtaoguo) [reply src/maxtext/integration/vllm/torchax_converter/gemma4_moe.py:297]: Removed

**Review by khatwanimohit — APPROVED** [review]
